app/models/mixins/inter_region_api_method_relay.rb

```diff
@@ -0,0 +1,80 @@
+module InterRegionApiMethodRelay
+  def self.extended(klass)
+    unless klass.const_defined?("InstanceMethodRelay")
+      instance_relay = klass.const_set("InstanceMethodRelay", Module.new)
+      klass.prepend(instance_relay)
+    end
+
+    unless klass.const_defined?("ClassMethodRelay")
+      class_relay = klass.const_set("ClassMethodRelay", Module.new)
+      klass.singleton_class.prepend(class_relay)
+    end
+  end
+
+  def api_relay_method(method, action = method)
+    relay = const_get("InstanceMethodRelay")
+    collection_name = collection_for_class
+
+    relay.class_eval do
+      define_method(method) do |*meth_args, &meth_block|
+        api_args = block_given? ? yield(*meth_args) : {}
+
+        if in_current_region?
+          super(*meth_args, &meth_block)
+        else
+          api_client = InterRegionApiMethodRelay.api_client_connection_for_region(region_number)
+          collection = api_client.send(collection_name)
+          obj = collection.find(id)
+          obj.public_send(action, api_args)
+        end
+      end
+    end
+  end
+
+  def api_relay_class_method(method, action = method)
+    relay = const_get("ClassMethodRelay")
+    collection_name = collection_for_class
+    raise ArgumentError, "A block is required to determine target object region and API arguments" unless block_given?
+
+    relay.class_eval do
+      define_method(method) do |*meth_args, &meth_block|
+        record_id, api_args = yield(*meth_args)
+
+        if id_in_current_region?(record_id)
+          super(*meth_args, &meth_block)
+        else
+          api_client = InterRegionApiMethodRelay.api_client_connection_for_region(id_to_region(record_id))
+          collection = api_client.send(collection_name)
+          collection.public_send(action, api_args || {})
```

Comment:
I'm not concerned about whether to use a block or not. I'm curious about what the best way to generalize use of the arguments is.

For example as @Fryguy mentioned [here](https://github.com/ManageIQ/manageiq/pull/11782#discussion_r82478503) should the args be passed in as one parameter (presumably a big hash) or should we splat and pass them separately? Do different calls use different formats?

It seems that the api-client gem raises if the arguments are not a hash in the case of resource actions, but is not doing the same thing (as far as I can tell, https://github.com/ManageIQ/manageiq-api-client/blob/master/lib/manageiq/api/client/collection.rb#L155-L188 is a bit dense for me) in the case of actions on collections.